File: examples/consultant/consultant.cpp
/****************************************************************************
**
** Copyright (C) 2002-2014 Integrated Computer Solutions, Inc.
** All rights reserved.
**
** This file is part of an example program for QicsTable. This example
** program may be used, distributed and modified without limitation.
**
*****************************************************************************/

#include "consultant.h"

#include <QFileDialog>
#include <QApplication>
#include <QLabel>
#include <QMenuBar>
#include <QMessageBox>

#include <QicsColumnHeader.h>
#include <QicsDateTimeCellDisplay.h>
#include <QicsCell.h>
#include <QicsColumn.h>
#include <QicsDataItemFormatter.h>
#include <QicsTable.h>
#include <QicsDataModelDefault.h>
#include <QicsComboCellDisplay.h>
#include <QicsCheckCellDisplay.h>

const int NUM_COLUMNS = 7;

/*
* SSNValidator Methods
*/

QValidator::State SSNValidator::validate( QString& text, int& pos ) const
{
    /* An empty string could be made acceptable */
    if(text.isNull())
        return Intermediate;

    /* Protect against spurious calls to validate() */
    if( pos > text.length() )
        return Intermediate;

    /* Anything but decimal digits and dashes is invalid. We only need
    * to check the character at the cursor positions. This speeds
    * things up massively.
    */
    if( ( pos-1 >= 0 ) && !text[pos-1].isDigit() && ( text[pos-1] != '-' ) )
        return Invalid;

    /* If the characters entered so far are valid, but the string
    * contains less than ten digits, it could be made acceptable, but
    * is not yet.
    */
    int numDigits = text.length();
    int numDashes = 0;

    for (int i=0; i < text.length(); ++i) {
        if(text.at(i) == QChar('-')) {
            --numDigits;
            ++numDashes;
        }
    }

    if( numDigits < 9 )
        return Intermediate;

    /* More than nine digits is always invalid */
    if( numDigits > 9 )
        return Invalid;

    if (numDashes != 0)
        return Intermediate;

    /* numDashes = 0 and numDigits = 9 */
    return Acceptable;
}

void SSNValidator::fixup( QString& text ) const
{
    QString inString = text;
    int inLength = inString.length();

    for ( int i=inLength; i > 0; --i) {
        if( inString[i] == '-' ) {
            text.remove(i,1);
            qDebug ("Removing dash. \n");
        }
    }
}

////////////////////////////////////////////////////////////////////////

/*
* The main Consult class
*/

Consultant::Consultant(QWidget *parent)
    : QMainWindow(parent)
{
    setAttribute(Qt::WA_DeleteOnClose);
    setWindowTitle(tr("Consultant Example"));

    // create the data model
    m_dataModel = new QicsDataModelDefault(0,NUM_COLUMNS);

    // create the table, using the data model we created above
    m_table = new QicsTable(m_dataModel, this);

    setCentralWidget(m_table);

    // customize the column header to make things look pretty
    m_table->columnHeaderRef().setAlignment(Qt::AlignCenter);

    m_table->columnHeaderRef().cellRef(0,0).setLabel("First Name");
    m_table->columnHeaderRef().cellRef(0,1).setLabel("Last Name");
    m_table->columnHeaderRef().cellRef(0,2).setLabel("S.S. Number");
    m_table->columnHeaderRef().cellRef(0,3).setLabel("Date of Birth");
    m_table->columnHeaderRef().cellRef(0,4).setLabel("Office");
    m_table->columnHeaderRef().cellRef(0,5).setLabel("Sec. Clearance");
    m_table->columnHeaderRef().cellRef(0,6).setLabel("Billable Rate");

    //Add a validator to the SSN to insure a consistant display
    //behaviour and error checking.
    SSNValidator* ssnValidator = new SSNValidator();
    m_table->columnRef(2).setValidator(ssnValidator);

    //Format the combo box for the office column
    QicsComboCellDisplay *officeComboDisplay = new QicsComboCellDisplay(m_table);
    officeComboDisplay->addItem("Boston");
    officeComboDisplay->addItem("Cambridge");
    officeComboDisplay->addItem("Helsinki");
    officeComboDisplay->setAddValueToList(false);
    m_table->columnRef(4).setDisplayer(officeComboDisplay);

    // Security clearance check box
    QicsCheckCellDisplay *columnHeaderCheckDisp = new QicsCheckCellDisplay(m_table);
    m_table->columnRef(5).setDisplayer(columnHeaderCheckDisp);

    // Creating date time displayer
    QicsDateTimeCellDisplay* dt = new QicsDateTimeCellDisplay(m_table);
    dt->setDisplayFormat("dd/MM/yyyy");

    m_table->columnRef(3).setDisplayer(dt);

    //Format the billable rate column with a dollar sign, decimal point,
    //and the appropriate cents digits.
    QicsDataItemSprintfFormatter *brFormatter = new QicsDataItemSprintfFormatter();
    brFormatter->addFormatString(QicsDataItem_Float, "$%.2f");
    brFormatter->addFormatString(QicsDataItem_Int, "$%d.00");
    m_table->columnRef(6).setFormatter(brFormatter);
    m_table->columnRef(6).setValidator(new QDoubleValidator(m_table));

    // Add a title widget to the top of the table
    QLabel *label = new QLabel(tr("Consultants"), m_table);
    label->setAlignment(Qt::AlignCenter);
    m_table->setTopTitleWidget(label);

    //Set column widths
    for (int i = 0; i < NUM_COLUMNS; ++i)
        m_table->columnRef(i).setWidthInChars(7);

    m_table->setVisibleColumns(NUM_COLUMNS);

    //Menu Bar
    QMenu *filemenu = new QMenu(tr("File"),this);

    QAction *save = filemenu->addAction(tr("Save..."));
    connect(save,SIGNAL(triggered()),this,SLOT ( fileSave()));

    QAction *open = filemenu->addAction(tr("Open..."));
    connect(open,SIGNAL(triggered()),this,SLOT ( fileOpen()));

    QAction *quit = filemenu->addAction( tr("Quit") );
    connect(quit,SIGNAL(triggered()),qApp,SLOT(quit()));

    QMenu *toolsmenu = new QMenu("Tools",this);

    QAction *addConsultant = toolsmenu->addAction(tr("Add Consultant"));
    connect(addConsultant,SIGNAL(triggered()),this,SLOT(addRow()));

    QAction *deleteConsultant = toolsmenu->addAction(tr("Delete Consultant"));
    connect(deleteConsultant,SIGNAL(triggered()),this,SLOT(deleteRow()));

    QAction *sortAscending = toolsmenu->addAction(tr("Sort Ascending"));
    connect(sortAscending,SIGNAL(triggered()),this,SLOT(sortAscending()));

    QAction *sortDescending = toolsmenu->addAction(tr("Sort Descending"));
    connect(sortDescending,SIGNAL(triggered()),this,SLOT(sortDescending()));

    // Help
    QMenu* help = new QMenu(tr("&Help"), this);
    help->addAction(QIcon(":/Resources/icon16_help.png"), tr("About..."), this, SLOT(about()), Qt::CTRL+Qt::Key_H);
    menuBar()->addMenu(help);

    QMenuBar *menubar = menuBar();
    menubar->addMenu(filemenu);
    menubar->addMenu(toolsmenu);
    menubar->addMenu(help);

    //Connection for autosort
    connect(&(m_table->columnHeaderRef()), SIGNAL(doubleClicked(int, int, int, const QPoint&)),
        this, SLOT(autoSort(int, int, int, const QPoint&)));

    //Reset autoSort sanity vars
    m_autoSortCol = -1;
    m_autoSortOrder = Qics::Descending;
}

void Consultant::fileSave()
{
    QString outFileName = QFileDialog::getSaveFileName(this, tr("Choose a file"), "./",
        tr("Serialized data files (*.dat)"));

    QFile file(outFileName);

    if(!file.open(QIODevice::WriteOnly)) {
        qWarning("error opening file for writing...");
        return;
    }

    QDataStream ds(&file);

    int nrows = m_table->dataModel()->numRows();
    int ncols = m_table->dataModel()->numColumns();

    ds << nrows << ncols;

    for(int i = 0;i < nrows; ++i) {
        for(int j = 0;j < ncols; ++j) {
            const QicsDataItem* itm = m_table->dataModel()->item(i,j);
            if(itm)
                itm->encode(ds);
            else {
                // if item is NULL creating empty string and serializing it!
                QicsDataString str;
                str.encode(ds);
            }
        }
    }
    file.close();
}

void Consultant::fileOpen()
{

    QString inFileName = QFileDialog::getOpenFileName(this, tr("Choose a filename"), "./",
        tr("Serialized data files (*.dat)"));

    if (inFileName.isEmpty())
        return;

    QFile file(inFileName);

    if(!file.open(QIODevice::ReadOnly)) {
        qWarning("error opening file for reading...");
        return;
    }

    QDataStream ds(&file);

    int rows;
    int columns;

    ds>>rows;
    ds>>columns;

    m_table->setIgnoreModelSizeChanges(true);
    m_table->setRepaintBehavior(Qics::RepaintOff);

    m_table->dataModel()->clearModel();
    m_table->dataModel()->addRows(rows);
    m_table->dataModel()->addColumns(columns);

    m_table->clearSelectionList();
    for(int i = 0;i<rows;i++) {
        for(int j = 0;j<columns;j++) {
            QicsDataItem *it = QicsDataItem::decode(ds);
            if(it)
                m_table->dataModel()->setItem(i,j, *it);
        }
    }
    m_table->setIgnoreModelSizeChanges(false);
    m_table->setRepaintBehavior(Qics::RepaintOn);

    file.close();
}

void Consultant::addRow()
{
    m_table->addRows(1);
}

void Consultant::deleteRow()
{
    m_table->deleteRow(m_table->currentCell()->rowIndex());
}

void Consultant::sortAscending()
{
    m_table->sortRows(m_table->currentCell()->columnIndex(), Qics::Ascending);
}

void Consultant::sortDescending()
{
    m_table->sortRows(m_table->currentCell()->columnIndex(), Qics::Descending);
}

void Consultant::autoSort(int , int col, int , const QPoint &)
{
    if (m_autoSortCol == col) {
        if (m_autoSortOrder == Qics::Ascending)
            m_autoSortOrder = Qics::Descending;
        else
            m_autoSortOrder = Qics::Ascending;
    }
    else {
        m_autoSortCol = col;
        m_autoSortOrder = Qics::Ascending;
    }

    m_table->sortRows(col, m_autoSortOrder);
}

void Consultant::about()
{
    // init helpDialog
    QMessageBox helpDialog(this);
    helpDialog.setText(
        "<table>"
        "<tr><td>"
        "<img src=':/Resources/ics.png'>"
        "<br><h3>Integrated Computer Solutions Incorporated</h3>"
        "</td><td>"
        "<h3>The User Interface Company</h3>"
        "54B Middlesex Turnpike<br>Bedford, MA USA 01730<br><br>"
        "<b>info@ics.com<br>www.ics.com</b>"
        "</td><tr><td colspan=2>"
        "<hr>"
        "This example represents Consultant based on QicsTable."
        "<p>"
        "<b>HOW TO USE</b><br>"
        "TOOD: Need user instruction"
        "</td></table>"
        );
    helpDialog.setWindowTitle( tr("About Consultant Example") );
    helpDialog.setIcon( QMessageBox::NoIcon );
    helpDialog.exec();
}

File: addons/pyqicstable/examples/consultant/consultant.py
#!/usr/bin/env python

## Copyright ( C ) 2002-2014 Integrated Computer Solutions, Inc.
## All rights reserved.
##
## This file is part of an example program for PyQicsTable.  This example
## program may be used, distributed and modified without limitation.

########################################################################
##
## This example program demonstrates the use of a number of QicsTable
## features, including custom cell displayers, validators, and formatters.
##
## The program displays a table when can be loaded with data concerning
## "consultants".  Data can be loaded from, and saved to, a text file.
## Columns in the table are:
##
## First Name:         (string) default cell displayer
## Last Name:          (string) default cell displayer
## SSN:                (int) default cell displayer, custom validator
##                           and formatter
## Date of Birth:      (date) default cell displayer
## Office:             (string) combo cell displayer
## Security Clearance: (string) subclassed check box displayer
## Billable Rate:      (float) default cell displayer, custom formatter
##
########################################################################


import sys


from PyQt4 import QtCore, QtGui, Qt

from qicstable import QicsTable, QicsDataModelDefault, QicsCheckCellDisplay
from qicstable import QicsComboCellDisplay, QicsDataItemSprintfFormatter
from qicstable import QicsDataItem_Float, QicsDataItem_Int, Qics, QicsDataString


NUM_COLUMNS = 7


class SecCheckBox(QicsCheckCellDisplay):
    """
    SecCheckBox - A subclass of QicsCheckCellDisplay that displays a "check" if
    the underlying data value is 'y'.  This class also stores (in the data
    model) any changes to the checkbox as 'y' if the box is checked, or 'n'
    otherwise.
    """
    def checkState(self, info, model_row, model_col, itm):
        """
        Reimplemented from QicsCheckCellDisplay.
        """
        # Returns the state of the cell, based on the data item. State is false
        # unless the data item is "y".
        if itm:
            tableState = itm.string()

            return (tableState.toLatin1() == "y")
        else:
            return False

    def valueChanged(self, info, model_row, model_col, set):
        """
        Reimplemented from QicsCheckCellDisplay.
        """
        if set:
            val = "y"
        else:
            val = "n"

        info.dataModel().setItem(model_row, model_col, QicsDataString(val))


class SSNValidator(QtGui.QValidator):
    """
    SSNValidator - A subclass of QValidator that restricts input to numeral and
    '-' chars.   Once the user presses enter, fixup() will strip the dashed
    from the input in order to keep the data consistant.  A formatter will be
    used to display the value in the table in the common manner (XXX-XX-XXXX).
    """
    def __init__(self):
        QtGui.QValidator.__init__(self, None)

    def validate(self, text, pos):
        # An empty string could be made acceptable
        if text.isEmpty():
            return (QtGui.QValidator.Acceptable, pos)

        # Protect against spurious calls to validate()
        if pos > text.length():
            return (QtGui.QValidator.Acceptable, pos)

        # Anything but decimal digits and dashes is invalid. We only need to
        # check the character at the cursor positions. This speeds things up
        # massively.

        if not text.at(pos-1).isDigit() and text[pos-1] != '-' and pos-1 >= 0:
            return (QtGui.QValidator.Invalid, pos)

        # If the characters entered so far are valid, but the string contains
        # less than ten digits, it could be made acceptable, but is not yet.
        numDigits = text.length()
        numDashes = 0

        for i in range(text.length()):
            if text[i] == '-':
                numDigits -= 1
                numDashes += 1

        if numDigits < 9:
            return (QtGui.QValidator.Acceptable, pos)

        # More than nine digits is always invalid
        if numDigits > 9:
            return (QtGui.QValidator.Invalid, pos)

        if numDashes != 0:
            return (QtGui.QValidator.Acceptable, pos)

        # numDashes = 0 and numDigits = 9
        return (QtGui.QValidator.Acceptable, pos)

    def fixup(self, text):
        inString = QtCore.QString(text)
        inLength = inString.length()

        for i in range(inLength, 0, -1):
            if inString[i] == '-':
                text.remove(i, 1)


class Consultant(QtGui.QMainWindow):
    """
    Consultant - The primary class.  Subclassed from QMainWindow, this class
    contains both the table and the data model, as well as all the methods to
    implement the UI behavior.
    """
    def __init__(self, parent, name):
        QtGui.QMainWindow.__init__(self, parent)

        # create the data model
        self.dataModel = QicsDataModelDefault(0, NUM_COLUMNS)

        # create the table, using the data model we created above
        self.table = QicsTable(self.dataModel, self)
        self.setCentralWidget(self.table)

        # customize the column header to make things look pretty
        self.table.columnHeaderRef().setAlignment(QtCore.Qt.AlignCenter)

        self.table.columnHeaderRef().cellRef(0,0).setLabel("First Name")
        self.table.columnHeaderRef().cellRef(0,1).setLabel("Last Name")
        self.table.columnHeaderRef().cellRef(0,2).setLabel("S.S. Number")
        self.table.columnHeaderRef().cellRef(0,3).setLabel("Date of Birth")
        self.table.columnHeaderRef().cellRef(0,4).setLabel("Office")
        self.table.columnHeaderRef().cellRef(0,5).setLabel("Sec. Clearance")
        self.table.columnHeaderRef().cellRef(0,6).setLabel("Billable Rate")

        # Add a validator to the SSN to insure a consistant display behaviour
        # and error checking.
        self.ssnValidator = SSNValidator()
        self.table.columnRef(2).setValidator(self.ssnValidator)

        # Format the combo box for the office column
        self.officeComboDisplay = QicsComboCellDisplay()
        self.officeComboDisplay.addItem("Boston")
        self.officeComboDisplay.addItem("Cambridge")
        self.officeComboDisplay.addItem("Helsinki")
        self.officeComboDisplay.setAddValueToList(False)
        self.table.columnRef(4).setDisplayer(self.officeComboDisplay)

        # Security clearance check box
        self.columnHeaderCheckDisp = SecCheckBox()
        self.table.columnRef(5).setDisplayer(self.columnHeaderCheckDisp)

        # Format the billable rate column with a dollar sign, decimal point,
        # and the appropriate cents digits.
        self.brFormatter = QicsDataItemSprintfFormatter()
        self.brFormatter.addFormatString(QicsDataItem_Float, "$%.2f")
        self.brFormatter.addFormatString(QicsDataItem_Int, "$%d.00")
        self.table.columnRef(6).setFormatter(self.brFormatter)
	self.table.columnRef(6).setValidator(QtGui.QDoubleValidator(self.table))

        # Add a title widget to the top of the table
        label = QtGui.QLabel("Consultants", self.table)
        label.setAlignment(QtCore.Qt.AlignCenter)
        self.table.setTopTitleWidget(label)

        # Set column widths
        for i in range(NUM_COLUMNS):
            self.table.columnRef(i).setWidthInChars(7)

        self.table.setVisibleColumns(NUM_COLUMNS)

        # Menu Bar
	menubar = self.menuBar()

	self.filemenu = QtGui.QMenu("File", self)

	self._Open = QtGui.QAction("Open...", self)
	self.filemenu.addAction(self._Open)
	self.connect(self._Open, QtCore.SIGNAL("triggered()"), self.fileOpen)

	self._Save = QtGui.QAction("Save...", self)
	self.filemenu.addAction(self._Save)
	self.connect(self._Save, QtCore.SIGNAL("triggered()"), self.fileSave)

	self._Exit = QtGui.QAction("Quit", self)
	self.filemenu.addAction(self._Exit)
	self.connect(self._Exit, QtCore.SIGNAL("triggered()"), QtGui.qApp, QtCore.SLOT("quit()"))

	menubar.addMenu(self.filemenu)

	self.toolsmenu = QtGui.QMenu("Tools", self)

	self._AddCs = QtGui.QAction("Add consultant", self)
	self._DelCs = QtGui.QAction("Delete consultant", self)
	self._SortA = QtGui.QAction("Sort Ascending", self)
	self._SortD = QtGui.QAction("Sort Descending", self)

	self.connect(self._AddCs, QtCore.SIGNAL("triggered()"), self.addRow)
	self.connect(self._DelCs, QtCore.SIGNAL("triggered()"), self.deleteRow)
	self.connect(self._SortA, QtCore.SIGNAL("triggered()"), self.sortAscending)
	self.connect(self._SortD, QtCore.SIGNAL("triggered()"), self.sortDescending)

	self.toolsmenu.addAction(self._AddCs)
	self.toolsmenu.addAction(self._DelCs)
	self.toolsmenu.addAction(self._SortA)
	self.toolsmenu.addAction(self._SortD)

	menubar.addMenu(self.toolsmenu)





        # Connection for autosort
        QtCore.QObject.connect(self.table.columnHeaderRef(),
                        QtCore.SIGNAL("doubleClicked(int, int, int, const QPoint &)"),
                        self.autoSort)

        # Reset autoSort sanity vars
        self._autoSortCol = -1
        self._autoSortOrder = Qics.Descending

    def addRow(self):
        self.table.addRows(1)

    def deleteRow(self):
        self.table.deleteRow(self.table.currentCell().rowIndex())

    def sortAscending(self):
        self.table.sortRows([self.table.currentCell().columnIndex()], Qics.Ascending)

    def sortDescending(self):
        self.table.sortRows([self.table.currentCell().columnIndex()], Qics.Descending)

    def fileOpen(self):
        inFileName = QtGui.QFileDialog.getOpenFileName(self, "Choose a file",
						 "./",
                                                 "CSV files (*.csv)")
						 #,
                                                 #"open file dialog")

        if inFileName.isEmpty():
            return;

        inFile = QtCore.QFile(inFileName)

        error = 0

        if inFile.open(QtCore.QIODevice.ReadOnly):
            inStream = QtCore.QTextStream(inFile)

            self.table.setIgnoreModelSizeChanges(True)
            self.dataModel.readASCII(inStream, ',', 0, 0, True)
            inFile.close()
            self.table.setIgnoreModelSizeChanges(False)
        else:
            error = 1

        if error:
            fileBox = QtGui.QMessageBox()
            fileBox.setCaption("File Operation")
            fileBox.setText(QtCore.QString("Could not read file: %1").arg(inFileName))
            fileBox.setButtonText(0, "OK")
            fileBox.exec_()

    def fileSave(self):
        outFileName = QtGui.QFileDialog.getSaveFileName(self, 
                                                  "save file dialog"
                                                  "Choose a file",
                                                  "./", 
                                                  "CSV files (*, csv)")

        if outFileName.isEmpty():
            return

        outFile = QtCore.QFile(outFileName)

        error = 0

        if outFile.open(QtCore.QIODevice.WriteOnly):
            outStream = QtCore.QTextStream(outFile)

            self.dataModel.writeASCII(outStream, ',')
            #outStream << "\n"
            outFile.close()
        else:
            error = 1

        fileBox = QtGui.QMessageBox()
        fileBox.setWindowTitle("File Operation")
        if error < 1:
            fileBox.setText(QtCore.QString("Saved Table Data to %1").arg(outFileName))
        else:
            fileBox.setText(QtCore.QString("Error Accessing %1").arg(outFileName))

        fileBox.setButtonText(0, "OK")
        fileBox.exec_()

    def autoSort(self, row, col, button, pos):
        if self._autoSortCol == col:
            if self._autoSortOrder == Qics.Ascending:
                self._autoSortOrder = Qics.Descending
            else:
                self._autoSortOrder = Qics.Ascending
        else:
            self._autoSortCol = col
            self._autoSortOrder = Qics.Ascending

        self.table.sortRows(col, self._autoSortOrder)


if __name__ == "__main__":
    app = QtGui.QApplication(sys.argv)

    consultant = Consultant(None, "Consultant Example")
    consultant.show()

    sys.exit(app.exec_())
